### PR TITLE
Perform bisection when verifying headers using the light client

### PR DESCRIPTION
**Pull request by romac**
Closes: #49

## Description

This PR implements proposal B described in #49.

All tasks spawned in the `start` command are now spawned on a local task set which executes them all on the same thread, ie. the main thread.

This seems like an acceptable solution to me given that the overall architecture of the relayer and the light client will likely change quite a bit in the coming weeks.
______

For contributor use:

- [ ] Unit tests written
- [ ] Added test to CI if applicable 
- [ ] Updated CHANGELOG_PENDING.md
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Comment by ancazamfir** [comment]
I tried to run the relayer on this branch and I don't see the light client update header messages. Just these:
`Apr 10 10:26:56.933  INFO relayer_cli::relayer: Relayer is running`

If I remove the relay thread then the relayer finishes quietly after `spawning light client...` messages
```
ancas-macbook-pro:1ibc-rs ancaz$ cargo run --bin relayer -- -c simple_config.toml start
...
Apr 10 10:30:21.657  INFO relayer_cli::commands::start: spawning light client chain.id=ibc0
Apr 10 10:30:21.658  INFO relayer_cli::commands::start: spawning light client chain.id=ibc1
ancas-macbook-pro:1ibc-rs ancaz$ 
```
Maybe I am doing smthg wrong, it's the first time I try this; followed the instructions from: https://github.com/informalsystems/ibc-rs/blob/master/README.md

**Comment by romac** [comment]
Yeah, I get the same output too. My bad, I should have mentioned that and marked the PR as a draft. Will let you know once(if) it works and is actually ready for review.

**Comment by ancazamfir** [comment]
ah ok, thought is ready for review :)

**Comment by romac** [comment]
I had forgotten to `.await` the client task 🤦‍♂️

This now works as expected but has not been rebased on master yet.

I rebased locally, but the `start` command now fails because of a deserialization error when fetching the `TrustOptions` from the store. Might be an issue with `tendermint-rs@v0.33`. Will investigate further!

**Comment by romac** [comment]
Blocked by https://github.com/informalsystems/tendermint-rs/pull/196#issuecomment-611969478.

**Comment by romac** [comment]
Unblocked now that https://github.com/informalsystems/tendermint-rs/pull/210 has been merged.

**Comment by romac** [comment]
@ancazamfir Ready for review.

File: relayer/cli/src/commands/start.rs [hunk 1]
```diff
@@ -27,60 +27,81 @@ impl Runnable for StartCmd {
...
+        info!(target: "relayer_cli::relayer", "relayer is running");
         interval.tick().await;
     }
 }
 
-async fn update_headers<C: Chain, S: Store<C>>(mut client: Client<C, S>) {
+async fn update_client<C: Chain, S: Store<C>>(mut client: Client<C, S>) {
     debug!(chain.id = %client.chain().id(), "updating headers");
 
     let mut interval = tokio::time::interval(Duration::from_secs(3));
```
Comment (ancazamfir) [inline relayer/cli/src/commands/start.rs:85]: From the documentation: "Note that when it starts, it produces item too."  So the  `interval.tick().await;` should be the first in the loop. I looked at the logs when running with two chains and noticed two updates per chain without the wait when starting the relayer.

File: relayer/cli/src/commands/start.rs [hunk 2]
```diff
@@ -27,60 +27,81 @@ impl Runnable for StartCmd {
...
 
     loop {
+        info!(chain.id = %client.chain().id(), "updating client");
```
Comment (ancazamfir) [inline relayer/cli/src/commands/start.rs:88]:
```suggestion
        interval.tick().await;
        info!(chain.id = %client.chain().id(), "updating client");
```
and remove from end of loop.

File: relayer/cli/src/commands/start.rs [hunk 3]
```diff
@@ -27,60 +27,81 @@ impl Runnable for StartCmd {
...
+
         let result = client.update(SystemTime::now()).await;
```
Comment (ancazamfir) [inline relayer/cli/src/commands/start.rs:90]: I believe this is the only place we should upate client, i.e. remove from `client.rs`
```
--- a/relayer/relay/src/client.rs
+++ b/relayer/relay/src/client.rs
@@ -93,10 +93,6 @@ where
             client.init_with_trust_options(trust_options).await?;
         }

-        // Perform an update already, to make sure the client is up-to-date.
-        // TODO: Should we leave this up to the responsibility of the caller?
-        let _ = client.update(SystemTime::now()).await?;
-
         Ok(client)
     }```

File: relayer/cli/src/commands/start.rs (hunk 1 above)
Comment (romac) [reply relayer/cli/src/commands/start.rs:85]: Good catch!

File: relayer/cli/src/commands/start.rs (hunk 3 above)
Comment (romac) [reply relayer/cli/src/commands/start.rs:90]: 👍

**Comment by romac** [comment]
Just addressed your comments :)

**Comment by ancazamfir** [comment]
> Just addressed your comments :)

Looks good! Thanks @romac!

**Pull request merged**